### PR TITLE
Implement authorization via google

### PR DESCRIPTION
**Pull request by vo333ua**
## JIRA

* https://ssu-jira.softserveinc.com/secure/RapidBoard.jspa?rapidView=933&view=detail&selectedIssue=PYTHON-144
* https://ssu-jira.softserveinc.com/secure/RapidBoard.jspa?rapidView=933&view=detail&selectedIssue=PYTHON-143
## Code reviewers

- [ ] @meelros
- [ ] @BoroviyOrest
- [ ] @Kardan1123
- [x] @mizin4ik
- [ ] @dimigor
- [ ] @PetrushynskyiOleksii
- [ ] @Sasha1152

### Second Level Review

- [ ] @lhalam

## Summary of change

Implement authorization via google

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: way_to_home/custom_user/views.py [hunk 1]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
         return HttpResponse('invalid credentials', status=400)
     login(request, user=user)
     return HttpResponse('operation was successful provided', status=200)
+
+
+@require_http_methods(["GET"])
+def auth_google(request):
+    """Function that provides getting url for confirmation access to user's data."""
+    google = OAuth2Session(client_id=CLIENT_ID, redirect_uri=REDIRECT_URL, state=STATE, scope=SCOPE)
```
Comment (BoroviyOrest) [inline way_to_home/custom_user/views.py]: REDIRECT_URL or REDIRECT_URI?

File: way_to_home/custom_user/views.py [hunk 2]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+    data = google.authorization_url(url=AUTH_URI, state=STATE)[0]
```
Comment (BoroviyOrest) [inline way_to_home/custom_user/views.py]: AUTH_URI or AUTH_URL?

File: way_to_home/custom_user/views.py [hunk 3]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+    if data:
+        return redirect(data)
+    return HttpResponse(status=400)
+
+
+@require_http_methods(["GET"])
+def signin_google(request):
+    """Function that provides user registration or authorization via google."""
+    google = OAuth2Session(client_id=CLIENT_ID, redirect_uri=REDIRECT_URL, state=STATE, scope=SCOPE)
```
Comment (BoroviyOrest) [inline way_to_home/custom_user/views.py]: Same as above

File: way_to_home/custom_user/views.py [hunk 4]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+    try:
+        code = request.GET["code"]
+    except ValueError:
+        return HttpResponse(status=400)
+    google.fetch_token(token_url=TOKEN_URI, client_secret=CLIENT_SECRET, code=code,
```
Comment (BoroviyOrest) [inline way_to_home/custom_user/views.py]: Same as above

File: way_to_home/custom_user/views.py [hunk 5]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+                       authorization_response='/api/v1/user/sign_in')
```
Comment (BoroviyOrest) [inline way_to_home/custom_user/views.py]: Use your path name

File: way_to_home/custom_user/views.py [hunk 6]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+    user_data = google.get('https://www.googleapis.com/oauth2/v1/userinfo').json()
+    if user_data:
+        user = CustomUser.get_by_email(user_data['email'])
+        if user:
+            login(request, user=user)
+            return HttpResponse(status=200)
+        user = CustomUser().create(email=user_data.get("email"), password='via_google')
+        user.save()
```
Comment (BoroviyOrest) [inline way_to_home/custom_user/views.py]: You don't need to call CustomUser constructor, also you don't need user.save(), user has been saved in CustomUser.create().

**Review by BoroviyOrest — DISMISSED** [review]

File: way_to_home/custom_user/views.py [hunk 7]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+        return HttpResponse(status=400)
+    google.fetch_token(token_url=TOKEN_URI, client_secret=CLIENT_SECRET, code=code,
+                       authorization_response='/api/v1/user/sign_in')
+    user_data = google.get('https://www.googleapis.com/oauth2/v1/userinfo').json()
+    if user_data:
+        user = CustomUser.get_by_email(user_data['email'])
+        if user:
+            login(request, user=user)
+            return HttpResponse(status=200)
+        user = CustomUser().create(email=user_data.get("email"), password='via_google')
```
Comment (BoroviyOrest) [inline way_to_home/custom_user/views.py]: Make hash from user's email

File: way_to_home/way_to_home/settings.py [hunk 8]
```diff
@@ -55,7 +55,7 @@
     'django.contrib.auth.middleware.AuthenticationMiddleware',
     'django.contrib.messages.middleware.MessageMiddleware',
     'django.middleware.clickjacking.XFrameOptionsMiddleware',
-    'middlewares.login_required.LoginRequiredMiddleware',
+    #'middlewares.login_required.LoginRequiredMiddleware',
```
Comment (BoroviyOrest) [inline way_to_home/way_to_home/settings.py]: Why did you comment this line?

File: way_to_home/custom_user/views.py [hunk 9]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+    except ValueError:
+        return HttpResponse(status=400)
+    google.fetch_token(token_url=TOKEN_URI, client_secret=CLIENT_SECRET, code=code,
+                       authorization_response='/api/v1/user/sign_in')
+    user_data = google.get('https://www.googleapis.com/oauth2/v1/userinfo').json()
+    if user_data:
+        user = CustomUser.get_by_email(user_data['email'])
+        if user:
+            login(request, user=user)
+            return HttpResponse(status=200)
```
Comment (dimigor) [inline way_to_home/custom_user/views.py]: send a success message to the client

File: way_to_home/custom_user/views.py [hunk 10]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+        user = CustomUser().create(email=user_data.get("email"), password='via_google')
+        user.save()
+        login(request, user=user)
+        return HttpResponse(status=201)
```
Comment (dimigor) [inline way_to_home/custom_user/views.py]: send a message with information

File: way_to_home/custom_user/views.py [hunk 11]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+
+    return HttpResponse(status=400)
```
Comment (dimigor) [inline way_to_home/custom_user/views.py]: Describe the response

File: way_to_home/way_to_home/settings.py (hunk 8 above)
Comment (dimigor) [reply way_to_home/way_to_home/settings.py]: Your code must be working with middlewares

File: way_to_home/custom_user/views.py [hunk 12]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+
+
+@require_http_methods(["GET"])
+def signin_google(request):
+    """Function that provides user registration or authorization via google."""
+    google = OAuth2Session(client_id=CLIENT_ID, redirect_uri=REDIRECT_URL, state=STATE, scope=SCOPE)
+    try:
+        code = request.GET["code"]
+    except ValueError:
+        return HttpResponse(status=400)
```
Comment (dimigor) [inline way_to_home/custom_user/views.py]: `return HttpResponse("Failed", status=400)`
return response with message that informs client what`s happend

File: way_to_home/custom_user/views.py [hunk 13]
```diff
@@ -63,3 +66,37 @@ def log_in(request):
...
+
+
+@require_http_methods(["GET"])
+def auth_google(request):
+    """Function that provides getting url for confirmation access to user's data."""
+    google = OAuth2Session(client_id=CLIENT_ID, redirect_uri=REDIRECT_URL, state=STATE, scope=SCOPE)
+    data = google.authorization_url(url=AUTH_URI, state=STATE)[0]
+    if data:
+        return redirect(data)
+    return HttpResponse(status=400)
```
Comment (dimigor) [inline way_to_home/custom_user/views.py]: message

File: way_to_home/way_to_home/settings.py (hunk 8 above)
Comment (vo333ua) [reply way_to_home/way_to_home/settings.py]: done

File: way_to_home/custom_user/views.py (hunk 7 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: done

File: way_to_home/custom_user/views.py (hunk 5 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: we havent name of this url

File: way_to_home/custom_user/views.py (hunk 3 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: done

File: way_to_home/custom_user/views.py (hunk 9 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: ok

File: way_to_home/custom_user/views.py (hunk 10 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: ok

File: way_to_home/custom_user/views.py (hunk 11 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: ok

File: way_to_home/custom_user/views.py (hunk 12 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: ok

File: way_to_home/custom_user/views.py (hunk 13 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: ok

File: way_to_home/custom_user/views.py [hunk 14]
```diff
@@ -63,3 +65,35 @@ def log_in(request):
         return HttpResponse('invalid credentials', status=400)
     login(request, user=user)
     return HttpResponse('operation was successful provided', status=200)
+
+
+@require_http_methods(["GET"])
+def auth_google(request):
+    """Function that provides getting url for confirmation access to user's data."""
+    google = OAuth2Session(client_id=CLIENT_ID, redirect_uri=REDIRECT_URI, state=STATE, scope=SCOPE)
```
Comment (MaxDubchak) [inline way_to_home/custom_user/views.py]: google is confusing variable name, consider google_session

File: way_to_home/custom_user/views.py [hunk 15]
```diff
@@ -63,3 +65,35 @@ def log_in(request):
...
+@require_http_methods(["GET"])
+def signin_google(request):
+    """Function that provides user registration or authorization via google."""
+    google = OAuth2Session(client_id=CLIENT_ID, redirect_uri=REDIRECT_URI, state=STATE, scope=SCOPE)
+    try:
+        code = request.GET["code"]
+    except ValueError:
+        return HttpResponse("Havent code", status=400)
+    google.fetch_token(token_url=TOKEN_URL, client_secret=CLIENT_SECRET, code=code)
+    user_data = google.get('https://www.googleapis.com/oauth2/v1/userinfo').json()
```
Comment (Sasha1152) [inline way_to_home/custom_user/views.py]: is this url have to be constant?

File: way_to_home/custom_user/views.py [hunk 16]
```diff
@@ -63,3 +65,35 @@ def log_in(request):
...
+
+
+@require_http_methods(["GET"])
+def signin_google(request):
+    """Function that provides user registration or authorization via google."""
+    google = OAuth2Session(client_id=CLIENT_ID, redirect_uri=REDIRECT_URI, state=STATE, scope=SCOPE)
+    try:
+        code = request.GET["code"]
+    except ValueError:
+        return HttpResponse("Havent code", status=400)
```
Comment (MaxDubchak) [inline way_to_home/custom_user/views.py]: ValueError is odd in this case, consider checking if code == None

Comment (MaxDubchak) [reply way_to_home/custom_user/views.py]: also "Havent code" is confusing
consider changing variable name of **code** too

File: way_to_home/way_to_home/settings.py [hunk 17]
```diff
@@ -130,6 +130,16 @@
 
 STATIC_URL = '/static/'
 
+# Required settings for authorization via google
+
+CLIENT_ID = '886654959108-v5nocnpcdjsgrovcthsb7o11uuf7g0am.apps.googleusercontent.com'
+CLIENT_SECRET = 'FwcIUXjKIzr4TEDmrXU4_7BS'
+REDIRECT_URI = 'http://localhost:8000/api/v1/user/signin_via_google'
```
Comment (Sasha1152) [inline way_to_home/way_to_home/settings.py:137]: Will this localhost always static?

**Review by MaxDubchak — DISMISSED** [review]

File: way_to_home/custom_user/views.py (hunk 16 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: done

File: way_to_home/custom_user/views.py (hunk 14 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: done

File: way_to_home/way_to_home/settings.py (hunk 17 above)
Comment (vo333ua) [reply way_to_home/way_to_home/settings.py:137]: Yes

File: way_to_home/custom_user/views.py (hunk 15 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: Yes

File: way_to_home/custom_user/views.py [hunk 18]
```diff
@@ -2,12 +2,14 @@
...
 from django.views.decorators.http import require_http_methods
+from django.shortcuts import redirect
+from requests_oauthlib import OAuth2Session
 
 from custom_user.models import CustomUser
 from utils.jwttoken import create_token, decode_token
 from utils.send_email import send_email
-from way_to_home.settings import DOMAIN
-
+from way_to_home.settings import (DOMAIN, CLIENT_ID, CLIENT_SECRET,
```
Comment (lhalam) [inline way_to_home/custom_user/views.py]: one element one line

File: way_to_home/way_to_home/settings.py [hunk 19]
```diff
@@ -130,6 +130,16 @@
 
 STATIC_URL = '/static/'
 
+# Required settings for authorization via google
+
+CLIENT_ID = '886654959108-v5nocnpcdjsgrovcthsb7o11uuf7g0am.apps.googleusercontent.com'
```
Comment (lhalam) [inline way_to_home/way_to_home/settings.py]: `CLIENT_ID = 'CLIENT_ID'`

Comment (vo333ua) [reply way_to_home/way_to_home/settings.py]: done

File: way_to_home/custom_user/views.py (hunk 18 above)
Comment (vo333ua) [reply way_to_home/custom_user/views.py]: done

**Review by lhalam — APPROVED** [review]

**Review by victoria-kit — APPROVED** [review]

**Review by dimigor — APPROVED** [review]

**Pull request merged**